File: frontend/src/types/goal.ts
export type GoalType = 'weekly' | 'monthly' | 'yearly'
export type GoalStatus = 'in_progress' | 'completed' | 'failed'

export interface Goal {
  id: number
  title: string
  description: string | null
  type: GoalType
  status: GoalStatus
  target_value: number
  current_value: number
  progress_percent: number
  deadline: string | null
  created_at: string
  updated_at: string
}

export interface GoalPayload {
  title: string
  description?: string | null
  type?: GoalType
  target_value?: number
  deadline?: string | null
}

export interface GoalUpdatePayload extends Partial<GoalPayload> {
  status?: GoalStatus
  current_value?: number
}

File: backend/app/schemas/goal.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import GoalStatus, GoalType


class GoalCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    type: GoalType = GoalType.weekly
    target_value: float = Field(default=1.0, gt=0)
    deadline: date | None = None


class GoalUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    type: GoalType | None = None
    status: GoalStatus | None = None
    target_value: float | None = Field(default=None, gt=0)
    current_value: float | None = Field(default=None, ge=0)
    deadline: date | None = None


class GoalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    type: GoalType
    status: GoalStatus
    target_value: float
    current_value: float
    progress_percent: float
    deadline: date | None
    created_at: datetime
    updated_at: datetime
